### PR TITLE
Recover bounded capture deadlines and preserve terminal session loss

### PR DESCRIPTION
**Pull request by miguelg719**
Bounds capture and RPC/batch waits. Two consecutive capture deadlines are recoverable; a third latches terminal session loss, while a successful tool call resets the counter. Late completion cannot overwrite newer snapshot IDs.

Terminal loss rejects queued calls without dispatch. Actions are never automatically replayed. Regressions cover bounded timeout inputs, late completion, counter reset, screenshot transport and first-loss handling.

Preserves provider/session/age/timeout diagnostics through the host, checks screenshot byte limits before decoding, and isolates failures in diagnostic observers. The typed RPC timeout error lives in a focused public module, keeping internal RPC helpers outside the SDK public-field contract.

Validation: 120 focused tests and 47 SDK AST/release guards, SDK/core/eval typechecks, SDK build/publint, and merged ownership/capture regression checks.

Stack position: 10/11. Base: `evals/consolidation-11-verifier-results`. Each PR contains the change relative to its immediate predecessor.

Reviewer entry points:

- [`packages/integrations/core/src/facade/tools.ts`](https://github.com/browserbase/stagehand/blob/2bc5787608993eee91e47b5c85f948bf4907f396/packages/integrations/core/src/facade/tools.ts)
- [`packages/sdk-ts/src/batch.ts`](https://github.com/browserbase/stagehand/blob/2bc5787608993eee91e47b5c85f948bf4907f396/packages/sdk-ts/src/batch.ts)
- [`packages/integrations/core/tests/facade-tools.test.ts`](https://github.com/browserbase/stagehand/blob/2bc5787608993eee91e47b5c85f948bf4907f396/packages/integrations/core/tests/facade-tools.test.ts)

The CLI startup fix is now included in the parent (#2892), so this PR focuses on capture/RPC deadlines and recovery. Current main is incorporated, preserving both the timeout exports and the runtime-compatibility exports.

Reordered stack: #2891 → #2892 → #2895 → #2896 → #2897 → #2898 → #2899 → #2900 → #2901 → #2893 → #2894.

Readiness: [CI passed](https://github.com/browserbase/stagehand/actions/runs/35200728466) on `2bc5787608`. Review threads are triaged and resolved.

### CODE REVIEW

File: packages/sdk-ts/src/rpcClient.ts [hunk 1]
```diff
@@ -54,8 +54,22 @@ type RegisteredRequestHandler = {
...
+  /** Replaces the method's derived response deadline for this one request. */
+  responseTimeoutMs?: number;
 };
 
+export class RPCResponseTimeoutError extends Error {
+  readonly method: string;
+  readonly timeoutMs: number;
+
+  constructor(method: string, timeoutMs: number) {
+    super(`RPC response timed out: ${method}`, { cause: { method, timeoutMs } });
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/sdk-ts/src/rpcClient.ts]: The typed RPC timeout now includes timeoutMs in its diagnostic while retaining the original message prefix. The error has a focused public module so internal RPC helpers remain outside the public SDK guard.

Fixed in `e043e1efb`, now merged with ownership fixes. Focused regressions, strict SDK AST guards and package typechecks pass.

File: packages/integrations/core/src/facade/stdio-server.ts [hunk 2]
```diff
@@ -136,11 +137,31 @@ async function createResources(): Promise<FacadeResources> {
     config.browser.type === "browserbase"
       ? await browserbase.launch(config.browser.launchOptions)
       : await localBrowser.launch(config.browser.launchOptions);
+  const launchedAt = Date.now();
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/integrations/core/src/facade/stdio-server.ts]: Session age now starts before the browser launch, so launch time is included. A fake-clock regression reproduces the previous undercount.

Fixed in `e043e1efb`, now merged with ownership fixes. Focused regressions, strict SDK AST guards and package typechecks pass.

File: packages/integrations/core/src/facade/stdio-server.ts [hunk 3]
```diff
@@ -136,11 +137,31 @@ async function createResources(): Promise<FacadeResources> {
...
+      // sessionAgeMs against the configured session timeout tells a Browserbase
+      // TIMED_OUT apart from a remote close.
+      onSessionLost: (loss) =>
+        process.stderr.write(
+          `${SESSION_LOST_TELEMETRY_PREFIX}${JSON.stringify({
+            ...loss,
+            cause: sanitizeErrorMessage(loss.cause),
+            provider: browser.provider,
+            ...(browser.sessionId && { sessionId: browser.sessionId }),
+            sessionAgeMs: Date.now() - launchedAt,
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/integrations/core/src/facade/stdio-server.ts:158]: Provider, session ID, age and configured timeout now survive the host parser with bounded numeric validation and sanitization. Regression tests cover accepted and malformed metadata.

Fixed in `e043e1efb`, now merged with ownership fixes. Focused regressions, strict SDK AST guards and package typechecks pass.

File: packages/integrations/core/src/facade/screenshot-transport.ts [hunk 4]
```diff
@@ -27,24 +27,64 @@ export function screenshotBase64BudgetFromArgs(args: string[]): number | undefin
...
   maxBase64Bytes: number,
+  maxSidePx = MAX_SCREENSHOT_SIDE_PX,
 ): Promise<TransportSafeScreenshot> {
   const attempts = screenshotAttempts(requested);
   for (const [index, options] of attempts.entries()) {
     const image = await capture(options);
-    if (Buffer.byteLength(image.data, "utf8") <= maxBase64Bytes) {
+    const size = imageDimensions(image);
+    const tooLarge = size !== undefined && Math.max(size.width, size.height) > maxSidePx;
+    if (!tooLarge && Buffer.byteLength(image.data, "utf8") <= maxBase64Bytes) {
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/integrations/core/src/facade/screenshot-transport.ts]: Byte-budget rejection now happens before image dimension decoding; oversized candidates are skipped without allocating their decoded buffer. The transport regression asserts this ordering.

Fixed in `e043e1efb`, now merged with ownership fixes. Focused regressions, strict SDK AST guards and package typechecks pass.

File: packages/integrations/core/src/facade/tools.ts [hunk 5]
```diff
@@ -322,12 +364,35 @@ export class StagehandFacadeTools {
...
+        // Permit two consecutive capture timeouts. The deadline does not cancel
+        // the underlying RPC, and recovery never replays that capture or an action.
+        if (error instanceof FacadeDeadlineError) {
+          this.consecutiveDeadlines += 1;
+          if (this.consecutiveDeadlines < StagehandFacadeTools.MAX_CONSECUTIVE_DEADLINES) {
+            throw error;
+          }
+          const cause = `executor unresponsive: ${this.consecutiveDeadlines} consecutive capture timeouts (last: ${error.message})`;
+          this.loss = { cause, tool, at: new Date().toISOString() };
+          this.options.onSessionLost?.(this.loss);
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/integrations/core/src/facade/tools.ts]: Synchronous throws and asynchronous rejections from onSessionLost are isolated. Both terminal paths preserve the first loss error and prevent subsequent dispatch, covered by regression tests.

Fixed in `e043e1efb`, now merged with ownership fixes. Focused regressions, strict SDK AST guards and package typechecks pass.

File: packages/evals/tui/commands/help.ts [hunk 6]
```diff
@@ -36,7 +34,9 @@ export function printHelp(): void {
   ]);
 }
 
-export function printRunHelp(): void {
+export async function printRunHelp(): Promise<void> {
+  const { listBenchHarnesses, listBenchHarnessesForTaskKind } =
+    await import("../../framework/benchHarness.js");
```
Comment (akeimach) [inline packages/evals/tui/commands/help.ts:39]: why lazy loading now?

**Review by akeimach — APPROVED** [review]

File: packages/evals/tui/commands/help.ts (hunk 6 above)
Comment (miguelg719) [reply packages/evals/tui/commands/help.ts:39]: we were importing every harness and sdk deps on `help` which took ~1.17s, so lazy loading gets us down to ~117ms
